Sources/PackageCollections/Storage/PackageCollectionsProfileStorage.swift

```diff
@@ -0,0 +1,339 @@
+    ///
+    /// - Parameters:
+    ///   - source: The `PackageCollectionSource` to add
+    ///   - order: Optional. The order that the source should take after being added to the profile.
+    ///            By default the new source is appended to the end (i.e., the least relevant order).
+    ///   - profile: The `Profile` to add source
+    ///   - callback: The closure to invoke when result becomes available
+    func add(source: PackageCollectionsModel.PackageCollectionSource,
+             order: Int?,
+             to profile: PackageCollectionsModel.Profile,
+             callback: @escaping (Result<Void, Error>) -> Void)
+
+    /// Removes source from the given profile.
+    ///
+    /// - Parameters:
+    ///   - source: The `PackageCollectionSource` to remove
+    ///   - profile: The `Profile` to remove source
+    ///   - callback: The closure to invoke when result becomes available
+    func remove(source: PackageCollectionsModel.PackageCollectionSource,
+                from profile: PackageCollectionsModel.Profile,
+                callback: @escaping (Result<Void, Error>) -> Void)
+
+    /// Removes source from the given profile.
+    ///
+    /// - Parameters:
+    ///   - source: The `PackageCollectionSource` to move
+    ///   - order: The order that the source should take in the profile.
+    ///   - profile: The `Profile` to move source
+    ///   - callback: The closure to invoke when result becomes available
+    func move(source: PackageCollectionsModel.PackageCollectionSource,
+              to order: Int,
+              in profile: PackageCollectionsModel.Profile,
+              callback: @escaping (Result<Void, Error>) -> Void)
+
+    /// Checks if a source is already in a profile.
+    ///
+    /// - Parameters:
+    ///   - source: The `PackageCollectionSource`
+    ///   - profile: Optional. The `Profile`, If not specified, checks across all profiles.
+    ///   - callback: The closure to invoke when result becomes available
+    func exists(source: PackageCollectionsModel.PackageCollectionSource,
+                in profile: PackageCollectionsModel.Profile?,
+                callback: @escaping (Result<Bool, Error>) -> Void)
+}
+
+// MARK: - FilePackageCollectionsProfileStorage
+
+struct FilePackageCollectionsProfileStorage: PackageCollectionsProfileStorage {
+    let fileSystem: FileSystem
+    let path: AbsolutePath
+
+    private let jsonEncoder: JSONEncoder
+    private let jsonDecoder: JSONDecoder
+
+    private let lock: LockProtocol
+    private let queue = DispatchQueue(label: "org.swift.swiftpm.FilePackageCollectionsProfileStorage")
+
+    init(fileSystem: FileSystem = localFileSystem, path: AbsolutePath? = nil) {
+        self.fileSystem = fileSystem
+
+        let name = "collections"
+        self.path = path ?? Paths.dotSwiftPM(fileSystem: fileSystem).appending(component: "\(name).json")
+
+        // FileLock only works with real FS
+        if fileSystem.isLocalFileSystem {
```

Comment:
Seems that this should work for all file systems, so clients don't have to have this kind of conditional.

Reply:
Didn't @tgymnich implement some solution to this as part of the work on the shared repository cache?

Reply:
```suggestion
        try fileSystem.withLock(on: cachePath.appending(component: name), type: .exclusive) {
```

Reply:
Yes, that's what I was thinking of.  It wouldn't need to check if it's a local file system, ideally.

Reply:
👀 ❤️